### PR TITLE
[ci] Docker tag library helper functions

### PR DESCRIPTION
**Pull request by khluu**
## Changes
- Add a library file for helper functions that work on Docker tags:
  - Get auth token from Docker Hub
  - Get auth token from Docker.io
  - Get tag's image creation time
  - Delete tag
  - Query tags from Docker Hub (with custom filter as input)
  - Query tags from Docker.io

## Checks
- Unit tests

### CODE REVIEW

**Comment by khluu** [comment]
@aslonnie Here is the workflow for the docker tags clean up. It's quite a long PR so please let me know if you'd like a few parts from the PR description to be moved to another PR.

**Comment by khluu** [comment]
@aslonnie  I have split the code so that this PR only contains helper functions.
#43309  is code change for deleting old commit tags
#43308  is for back up release tags
I have another pending branch for the binary file to run the task with Bazel. Will put up a PR for that after these PRs are merged in.

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    """
+    Retrieve Docker Registry token.
+
+    Args:
+        namespace: Docker namespace
+        repository: Docker repository
+
+    Returns:
+        Auth token for Docker.
+    """
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:51]: just remove the doc here, it is restating the function name.

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    )
+    auth_url = f"https://auth.docker.io/token?service={service}&scope={scope}"
+    response = requests.get(auth_url)
+    if response.status_code != 200:
+        raise AuthTokenException(f"Docker. Error code: {response.status_code}")
+    token = response.json().get("token", None)
+    return token
+
+
+def get_docker_hub_auth_token():
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:64]: send in user and password via args.

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+        "password": password,
+    }
+    headers = {"Content-Type": "application/json"}
+    response = requests.post(url, headers=headers, json=json_body)
+    if response.status_code != 200:
+        raise AuthTokenException(f"Docker Hub. Error code: {response.status_code}")
+    return response.json().get("token", None)
+
+
+def _get_git_log(n_days: int = 30):
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:83]: add return value type.

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+            f"--until='{n_days} days ago'",
+            "--pretty=format:%H",
+        ],
+        text=True,
+    )
+
+
+def list_recent_commit_short_shas(
+    n_days: int = 30,
+) -> List[str]:
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:97]: nit nit: can fit in one line?

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+
+    Args:
+        n_days: Number of days to go back in git log.
+
+    Returns:
+        List of recent commit SHAs (6 char).
+    """
+    commit_shas = _get_git_log(n_days=n_days)
+    short_commit_shas = [
+        commit_sha[:6] for commit_sha in commit_shas.split("\n") if commit_sha
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:109]: maybe define a const for `6`?

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+
+def _call_crane_config(tag: str):
+    try:
+        return subprocess.check_output(
+            ["crane", "config", tag],
+            stderr=subprocess.STDOUT,
+            text=True,
+        )
+    except subprocess.CalledProcessError as e:
+        return e.output
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:122]: maybe just call the OCI API to get the image config?

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+    repository, tag_name = repo_tag.split(":")
+
+    url = f"https://hub.docker.com/v2/repositories/{namespace}/{repository}/tags/{tag_name}"  # noqa E501
+    response = requests.delete(url, headers=headers)
+    if response.status_code == 429:
+        raise DockerHubRateLimitException()
+    if response.status_code != 204:
+        logger.info(f"Failed to delete {tag}, status code: {response.json()}")
+        return False
+    else:
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:165]: nit: remove the else

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        logger.info(f"Deleted tag {tag}")
+        return True
+
+
+def query_tags_from_docker_hub(
+    filter_func: Callable[[str], bool],
+    namespace: str,
+    repository: str,
+    num_tags: Optional[int] = 1000,
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:174]: set this to None for default. and if you want to set another default, you can `if num_tags is None: num_tags = 1000`

`num_tags = 0` or `-1` for no max, if there is a such a use case.

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+    Args:
+        namespace: Docker namespace
+        repository: Docker repository
+
+    Returns:
+        List of tags from Docker Registry in format namespace/repository:tag.
+    """
+    get_docker_auth_token(namespace=namespace, repository=repository)
+    result = _call_crane_ls(namespace=namespace, repository=repository)
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:256]: just write `get_docker_auth_token(namespace, repository)`

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+    try:
+        return subprocess.check_output(
+            [
+                "crane",
+                "ls",
+                f"{namespace}/{repository}",
+            ],
+            text=True,
+        )
+    except subprocess.CalledProcessError as e:
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:240]: call api directly?

Comment (khluu) [reply ci/ray_ci/automation/docker_tags_lib.py]: done

File: ci/ray_ci/automation/docker_tags_lib.py (hunk 9 above)
Comment (khluu) [reply ci/ray_ci/automation/docker_tags_lib.py]: done

File: ci/ray_ci/automation/docker_tags_lib.py (hunk 7 above)
Comment (khluu) [reply ci/ray_ci/automation/docker_tags_lib.py]: done

File: ci/ray_ci/automation/docker_tags_lib.py (hunk 6 above)
Comment (khluu) [reply ci/ray_ci/automation/docker_tags_lib.py]: done

File: ci/ray_ci/automation/docker_tags_lib.py (hunk 5 above)
Comment (khluu) [reply ci/ray_ci/automation/docker_tags_lib.py]: done

File: ci/ray_ci/automation/docker_tags_lib.py (hunk 3 above)
Comment (khluu) [reply ci/ray_ci/automation/docker_tags_lib.py]: done

File: ci/ray_ci/automation/docker_tags_lib.py (hunk 2 above)
Comment (khluu) [reply ci/ray_ci/automation/docker_tags_lib.py]: done

File: ci/ray_ci/automation/docker_tags_lib.py (hunk 1 above)
Comment (khluu) [reply ci/ray_ci/automation/docker_tags_lib.py]: done

File: ci/ray_ci/automation/docker_tags_lib.py (hunk 8 above)
Comment (khluu) [reply ci/ray_ci/automation/docker_tags_lib.py]: changed to `None`. For no max, I use an if statement `if num_tags` which should cover case where `num_tags=None` and `num_tags=0` (which means no max) don't apply.

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+class AuthTokenException(Exception):
+    """
+    Exception for failing to retrieve auth token.
+    """
+
+    def __init__(self, message: str):
+        super().__init__(f"Failed to retrieve auth token from {message}.")
+
+
+def get_docker_auth_token(namespace: str, repository: str) -> Optional[str]:
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:41]: we talked about making these private functions right? should they start with an underscore?

File: ci/ray_ci/automation/docker_tags_lib.py [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+def get_image_creation_time(tag: str) -> datetime:
+    """
+    Get Docker image creation time from tag image config.
+
+    Args:
+        tag: Docker tag name
+
+    Returns:
+        Datetime object of image creation time.
+    """
```
Comment (aslonnie) [inline ci/ray_ci/automation/docker_tags_lib.py:136]: and also remove the docs for functions that are obvious from the function name?

**Review by aslonnie — APPROVED** [review]

File: ci/ray_ci/automation/docker_tags_lib.py (hunk 11 above)
Comment (khluu) [reply ci/ray_ci/automation/docker_tags_lib.py]: Sorry I missed this while fixing the PR last time. Fixed now!

File: ci/ray_ci/automation/docker_tags_lib.py (hunk 12 above)
Comment (khluu) [reply ci/ray_ci/automation/docker_tags_lib.py]: done!

**Comment by aslonnie** [comment]
ok, I am merging this. nothing that requires blocking. we can also fix small things afterwards.

**Pull request merged**